### DISCUSSION TITLE
Discrepancy in actor forward direction between client and server while aiming and rotating in place

### DISCUSSION THREAD

**Discussion by jomath229** [Sixze/ALS-Refactored#541] [Q&A]
Discrepancy possible between server and client with rotate in place while in aiming mode. It is possible for the character on the server to not have the same orientation from the client. Since the rotation becomes driven by animation from the rotate in place anims and they get smoothed out with lerps etc., they can eventually fall out of sync and not have the same body orientation.

I implemented some proning logic and when implementing some collision depenetration, I realised the character's rotation on the server was out of sync since some rubberbanding occured on the client when some depenetration was detected from the body orientation which was different on the server, where the body orientation wasn't in a wall yet.

From what I see, it seems like actor forward direction between server and client can get out of sync while aiming and rotating in place, but it's hard to notice since it evens out when you start moving, but I've managed from jerking left to right to have a good 40/50 degrees of difference on the character orientation between both the client and the server, to the point that the client character started to rotate in place anim and the server character didn't rotate in place, leaving both of them at completely different rotations until you move again. Must I specify, I locked the rotation type to always be in aiming to follow a fortnish type of movement.

**Comment by Sixze** [comment]
Replication of character rotation to clients was intentionally disabled as it causes issues when the animation controls rotation, such as a severe jitter. In many other games, turns in place are usually implemented as fake turns, where the capsule's rotation is completely controlled by the CharacterMovementComponent, and animations modify the rotation of the root bone of the mesh to make it appear that the character is rotating smoothly under the influence of the animation. This approach can be seen in the Game Animation Sample, but it's not perfect either, it leads to leg sliding and abrupt motions.

Unfortunately this issue is not easy to fix. It is difficult to make it so that the rotations were fair and look good, and there were no issues in multiplayer, so you have to choose one of the two, and in my case I chose the first.

Maybe in the future this will change, but no promises.

**Reply by jomath229** [reply]
I ended up replicating the client's rotation value to the server, and used that one to calculate the proning depentration. I get where you come from, and since the capsule is generally always upright, having the same rotation between client and server is not that important and will end up not looking smooth indeed if it's used. On the server side, I simply use the actual client replicated yaw rotation as the forward vector when simulting collision etc. since I want to calculate the proning collision depending on where the legs are currrently on the client. and not vice-versa, since the legs could clip if the server is the one holding the rotation.

**Reply by Happier1945** [reply]
When will the project be upgraded to version 5.4 😭🙏 @Sixze
